### ISSUE TITLE
[css-overflow-5] Can we relax size containment on ::scroll-marker-group?

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#11166]
The [spec currently recommends](https://drafts.csswg.org/css-overflow-5/#scroll-marker-group-pseudo) the following default UA stylesheet:

```css
/* The generation of ::scroll-marker pseudo-elements shouldn’t
 * invalidate layout outside of this pseudo-element. */
::scroll-marker-group { contain: size !important; }
```

This is necessary to prevent cycles in certain cases, e.g. 
```html
<style>
.container { display: flex; height: 100vh; flex-direction: column; }
.scroller { columns: 1; flex: 1; scroll-marker-group: after; counter-reset: marker; }
.scroller::column::scroll-marker { counter-increment: marker; content: counter(marker); }
.scroller::scroll-marker-group { /* ... */ }
</style>
<div class="container"><div class="scroller">...</div></div>
```

Without size containment, the size of the `::scroll-marker-group` depends on the number / layout of the `::scroll-marker` elements that are within it. However, when the size of the group changes, it changes the space available for the scroller which could change the number of `::column`s.

However, having this size containment means that even in cases where the size doesn't affect the size of the scroller (e.g. `position: absolute`), they can't have it automatically sized and aligned based on its size, which adds a fair bit of complexity to achieving certain designs.

Some ideas that might make this less surprising or easier to work with:
1. Remove size containment and enforce out of flow positioning. This would make anchor positioning cases "just work", however would make it harder if you wanted to reserve space for the marker group to ensure it's not on top of anything.
2. Apply `overflow: auto` to make the problem more obvious when you're overflowing the space.
3. Something conditional? I'm not sure if there's any precedent for this.
4. Allow one layout cycle similar to a ResizeObserver based size change? This could be achieved by updating the initial `contain-intrinsic-size` after the first layout perhaps?
5. Something else?

**Comment by argyleink** [comment]
> something else

I could use anchor-size() when anchoring, for at least one of the sides, that could help with alignment but wouldnt be intrinsic sizing. but maybe worth noting here that anchor-size() can add a little, or in some scenarios, but wouldnt necessarily solve the unintrinsic nature of contain: size.

**Comment by argyleink** [comment]
just went to go put a background color on the `scroll-marker-group`, was like, where is it?! oh, no height/width, no background color.
<img width="264" alt="Screenshot 2024-11-07 at 1 00 27 PM" src="https://github.com/user-attachments/assets/7d8223a4-341c-4488-801f-7a888b304e98">

**Comment by argyleink** [comment]
also, having width on this marker group also means i can position buttons on either side of it with anchor nice and easy. without the width, the positioning gets tricky.

**Comment by flackr** [comment]
@mstensho Yep, that's option 3. I think that would be enough for many common use cases, I'm just not sure if:
1. There's any precedent for this sort of position-dependent containment,
2. If it wouldn't be better to try not to get in developers ways since most of the time even with in-flow examples the layout would be stable.

I could still see it being a common gotcha when you're using in-flow positioning, and potentially hard to know what size to set in these cases.

**Comment by mstensho** [comment]
> If it wouldn't be better to try not to get in developers ways since most of the time even with in-flow examples the layout would be stable.

In-flow without containment?

```html
<!DOCTYPE html>
<style>
  .container {
    width: 700px;
  }
  .scroller {
    overflow: auto;
    columns: 1;
    height: 110px;
    scroll-marker-group: before;
    background: yellow;
  }
  .scroller::scroll-marker-group {
    /* Look at me! No extrinsic size! */
    float: left;
    background: hotpink;
  }
  .item {
    display: inline-block;
    width: 300px;
    height: 70px;
    margin: 10px;
    background: cyan;
  }
  .scroller::column::scroll-marker {
    display: inline-block;
    width: 50px;
    height: 50px;
    content: "*";
  }
</style>
<div class="container">
  <div class="scroller">
    <div class="item"></div>
    <div class="item"></div>
    <div class="item"></div>
    <div class="item"></div>
    <div class="item"></div>
  </div>
</div>
```

With a `.container` width of 700px, two `.item`s (of width 300px plus margins and whitespace) will fit beside each other in each column. With 5 items, that means 3 columns, and therefore 3 scroll markers. 3 scroll markers need 50px * 3 plus some whitespace; let's say 160px. The scroll marker group is floated, and will be placed beside the scroller. If we were to take that into account, now there's only 700px - 160px = 540px space available for the scroller, which means that only one item will fit in each column (since items are 300px wide). Which means that we'll get 5 columns. And the final width of the scroll marker group should become 50px * 5 plus some whitespace; let's say 270px. 

So I guess we can say that it's stable in this case, although it took a couple of rounds. At least no circular dependencies (but I'm sure it's possible to come up with an example of that as well).

The main problem with supporting this (apart from detecting and speccing circular dependencies) is that we would have to do non-trivial amounts of work on every layout algorithm, basically teach them all that the contents of one element may affect the contents of a sibling. This would be a project comparable to adding support for block fragmentation to all layout algorithms, which we did, since we had to, and that took some years, and, like block fragmentation, it would add a heavy maintenance burden.

I'd like to go through the other options, the way I see them:

Option 1 isn't any weirder than what we currently have. Instead of enforced containment, we have enforced out-of-flow. Exchanging one disadvantage for another. I think enforcing containment is less of a disadvantage than enforcing out-of-flow, though?

Not sure what option 2 would give us, apart from annoying behavior if the developer has messed up. :)

Option 4 will allow for some chain reactions (example above) (probably even circular dependencies). The size of the scroll marker group could then affect the size of the scroller and/or its ancestors.

So option 3 seems like the best option to me, although implying containment from in-flowness is a new invention.

**Comment by flackr** [comment]
Yep, it is certainly possible to create some cycles if we had something generic like option 4. There are ways to mitigate it, e.g. similar to scrollbar logic where you only ever move forward and/or limit extra cycles per generated frame (could flicker between states like hover), however this potentially adds a lot of complexity and would still require one more layout pass in the common case.

As such, I'm inclined to agree that option 3 is probably the best way to get most of the use cases without all of this complexity.

**Comment by mstensho** [comment]
Dropping size containment on out-of-flow is fine, but I've come to realize that we always need layout containment. If a scroll-marker-group is out-of-flow, and it has scroll markers that are fixed-positioned, they would otherwise escape the scroll marker group, which would be bad, since the scroll-marker-group is populated (and depopulated) during layout.

**Comment by flackr** [comment]
Sounds good, i think that still allows for the expected use cases to work just fine. I made a demo to play with it:
https://codepen.io/flackr/pen/yyBZydb
